src/appleseed/renderer/modeling/bssrdf/randomwalkbssrdf.cpp

```diff
@@ -0,0 +1,408 @@
+            initial_dir = outgoing_point.get_shading_basis().transform_to_parent(initial_dir);
+            ShadingRay ray = ShadingRay(
+                outgoing_point.get_biased_point(initial_dir),
+                initial_dir,
+                outgoing_point.get_time(),
+                VisibilityFlags::ShadowRay,
+                outgoing_point.get_ray().m_depth + 1
+            );
+
+            ShadingPoint shading_points[2];
+            int next_shading_point_idx = 0;
+
+            DistanceSample distance_sample;
+            distance_sample.m_pivot = nullptr;
+            distance_sample.m_outgoing_point = &outgoing_point;
+            distance_sample.m_incoming_point = &shading_points[next_shading_point_idx];
+            distance_sample.m_incoming_point->clear();
+            distance_sample.m_volume_ray = &ray;
+            distance_sample.m_transmitted = false;
+
+            // Initialize BSSRDF value.
+            bssrdf_sample.m_value.set(1.0f);
+            bssrdf_sample.m_probability = 1.0f;
+
+            // Do random walk until we reach the surface from inside.
+            int n_iteration = 0;
+            const int MaxIterationsCount = 64;
+            while (!distance_sample.m_transmitted)
+            {
+                if (++n_iteration > MaxIterationsCount) break;
+
+                m_volume->sample_distance(
+                    shading_context,
+                    sampling_context,
+                    &values->m_volume_data,
+                    distance_sample);
+
+                assert(distance_sample.m_incoming_point->is_valid());
+
+                if (!distance_sample.m_transmitted)
+                {
+                    bssrdf_sample.m_value *= values->m_volume_data.m_scattering;
+                    bssrdf_sample.m_value /= values->m_volume_data.m_precomputed.m_extinction;
+
+                    // Find next random-walk direction using isotropic scattering.
+                    sampling_context.split_in_place(2, 1);
+                    const Vector2d s = sampling_context.next2<Vector2d>();
+                    ray = ShadingRay(
+                        shading_points[next_shading_point_idx].get_point(),
+                        sample_sphere_uniform(s),
+                        outgoing_point.get_time(),
+                        VisibilityFlags::ShadowRay,
+                        outgoing_point.get_ray().m_depth + 1
+                    );
+                    distance_sample.m_value.set(1.0f);
+                    distance_sample.m_volume_ray = &ray;
+
+                    // Update shading points.
+                    next_shading_point_idx = 1 - next_shading_point_idx;
+                    distance_sample.m_outgoing_point = distance_sample.m_incoming_point;
+                    distance_sample.m_incoming_point = &shading_points[next_shading_point_idx];
+                    distance_sample.m_incoming_point->clear();
+                }
+            }
+
+            if (!distance_sample.m_transmitted)
```

Comment:
I think it would be clearer (and it would allow avoiding one test) to move this into the

     if (++n_iteration > MaxIterationsCount)
     {
     }

test in the loop.